src/details/ArborX_DetailsSelfCollision.hpp

```diff
@@ -0,0 +1,186 @@
+/****************************************************************************
+ * Copyright (c) 2017-2022 by the ArborX authors                            *
+ * All rights reserved.                                                     *
+ *                                                                          *
+ * This file is part of the ArborX library. ArborX is                       *
+ * distributed under a BSD 3-clause license. For the licensing terms see    *
+ * the LICENSE file in the top-level directory.                             *
+ *                                                                          *
+ * SPDX-License-Identifier: BSD-3-Clause                                    *
+ ****************************************************************************/
+
+#ifndef ARBORX_DETAILS_SELF_COLLISION_HPP
+#define ARBORX_DETAILS_SELF_COLLISION_HPP
+
+#include <ArborX_AccessTraits.hpp>
+#include <ArborX_DetailsAlgorithms.hpp>
+#include <ArborX_DetailsHappyTreeFriends.hpp>
+#include <ArborX_DetailsHeap.hpp>
+#include <ArborX_DetailsKokkosExtArithmeticTraits.hpp>
+#include <ArborX_DetailsNode.hpp> // ROPE_SENTINEL
+#include <ArborX_DetailsOperatorFunctionObjects.hpp>
+#include <ArborX_LinearBVH.hpp>
+#include <ArborX_Predicates.hpp>
+#include <ArborX_Sphere.hpp>
+
+namespace ArborX::Details
+{
+
+template <class BVH, class Predicate, class Callback>
+struct SelfCollisionSpatial
+{
+  BVH _bvh;
+  Predicate _predicate;
+  Callback _callback;
+
+  template <class ExecutionSpace>
+  SelfCollisionSpatial(ExecutionSpace const &space, BVH const &bvh,
+                       Predicate const &predicate, Callback const &callback)
+      : _bvh{bvh}
+      , _predicate{predicate}
+      , _callback{callback}
+  {
+    if (_bvh.empty())
+    {
+      // do nothing
+    }
+    else if (_bvh.size() == 1)
+    {
+      // do nothing either
+    }
+    else
+    {
+      auto const n = _bvh.size();
+      Kokkos::parallel_for(
+          "ArborX::SelfCollision::spatial",
+          Kokkos::RangePolicy<ExecutionSpace>(space, n - 1, 2 * n - 1), *this);
+    }
+  }
+
+  KOKKOS_FUNCTION void operator()(int i) const
+  {
+    auto const predicate =
+        _predicate(HappyTreeFriends::getBoundingVolume(_bvh, i));
+    auto const leaf_permutation_i =
+        HappyTreeFriends::getLeafPermutationIndex(_bvh, i);
+
+    int node;
+#ifdef V1
+    int next = HappyTreeFriends::getRoot(_bvh); // start with root
+#else
+    int next = HappyTreeFriends::getRope(_bvh, i);
+#endif
+    do
+    {
+      node = next;
+
+      if (predicate(HappyTreeFriends::getBoundingVolume(_bvh, node)))
+      {
+        if (!HappyTreeFriends::isLeaf(_bvh, node))
+        {
+          auto const left_child = HappyTreeFriends::getLeftChild(_bvh, node);
+#ifdef V1
+          if (left_child < i)
+#else
+          if (true)
```

Comment:
One of the benefits of the V2 version that I didn't think about is that it is agnostic to the ordering of the internal nodes. Meaning that as long as the ropes are appropriately constructed, it will find correct answer. But I think V1 would only work with Karras ordering, and not Apetrei or any other.